src/traces/bar/defaults.js

```diff
@@ -61,22 +62,49 @@ function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
 function crossTraceDefaults(fullData, fullLayout) {
     var traceIn, traceOut;
 
-    function coerce(attr) {
-        return Lib.coerce(traceOut._input, traceOut, attributes, attr);
+    function coerce(attr, dflt) {
+        return Lib.coerce(traceOut._input, traceOut, attributes, attr, dflt);
     }
 
-    if(fullLayout.barmode === 'group') {
-        for(var i = 0; i < fullData.length; i++) {
-            traceOut = fullData[i];
+    for(var i = 0; i < fullData.length; i++) {
+        traceOut = fullData[i];
+
+        if(traceOut.type === 'bar') {
+            traceIn = traceOut._input;
+            // `marker.cornerradius` needs to be coerced here rather than in handleStyleDefaults()
+            // because it needs to happen after `layout.barcornerradius` has been coerced
+            var r = coerce('marker.cornerradius', fullLayout.barcornerradius);
+            traceOut.marker.cornerradius = validateCornerradius(r);
 
-            if(traceOut.type === 'bar') {
-                traceIn = traceOut._input;
+            if(fullLayout.barmode === 'group') {
                 handleGroupingDefaults(traceIn, traceOut, fullLayout, coerce);
             }
         }
     }
 }
 
+// Returns a value equivalent to the given cornerradius value, if valid;
+// otherwise returns`undefined`.
+// Valid cornerradius values must be either:
+//   - a numeric value (string or number) >= 0, or
+//   - a string consisting of a number >= 0 followed by a % sign
+// If the given cornerradius value is a numeric string, it will be converted
+// to a number.
+function validateCornerradius(r) {
+    var validR = false;
+    if(isNumeric(r)) {
+        r = +r;
+        if(r >= 0) validR = true;
```

Comment:
```suggestion
        if(r >= 0) return r;
```
don't really need to create the `validR` var, just return when you find a valid value (and if you get to the end, `return undefined;`)

Reply:
updated!